File: src/com/owncloud/android/ui/activity/PassCodeActivity.java
/**
 *   ownCloud Android client application
 *
 *   @author Bartek Przybylski
 *   @author masensio
 *   @author David A. Velasco
 *   Copyright (C) 2011 Bartek Przybylski
 *   Copyright (C) 2015 ownCloud Inc.
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License version 2,
 *   as published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.owncloud.android.ui.activity;

import java.util.Arrays;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.owncloud.android.R;
import com.owncloud.android.lib.common.utils.Log_OC;

public class PassCodeActivity extends AppCompatActivity {

    private static final String TAG = PassCodeActivity.class.getSimpleName();

    public final static String ACTION_ENABLE = PassCodeActivity.class.getCanonicalName() +
            ".ENABLE";
    public final static String ACTION_DISABLE = PassCodeActivity.class.getCanonicalName() +
            ".DISABLE";
    public final static String ACTION_REQUEST = PassCodeActivity.class.getCanonicalName()  +
            ".REQUEST";

    private Button mBCancel;
    private TextView mPassCodeHdr;
    private TextView mPassCodeHdrExplanation;
    private EditText[] mPassCodeEditTexts = new EditText[4];
    
    private String [] mPassCodeDigits = {"","","",""};
    private static String KEY_PASSCODE_DIGITS = "PASSCODE_DIGITS";
    private boolean mConfirmingPassCode = false;
    private static String KEY_CONFIRMING_PASSCODE = "CONFIRMING_PASSCODE";

    private boolean mBChange = true; // to control that only one blocks jump


    /**
     * Initializes the activity.
     *
     * An intent with a valid ACTION is expected; if none is found, an
     * {@link IllegalArgumentException} will be thrown.
     *
     * @param savedInstanceState    Previously saved state - irrelevant in this case
     */
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.passcodelock);
        
        mBCancel = (Button) findViewById(R.id.cancel);
        mPassCodeHdr = (TextView) findViewById(R.id.header);
        mPassCodeHdrExplanation = (TextView) findViewById(R.id.explanation);
        mPassCodeEditTexts[0] = (EditText) findViewById(R.id.txt0);
        mPassCodeEditTexts[0].requestFocus();
        getWindow().setSoftInputMode(
                android.view.WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE);
        mPassCodeEditTexts[1] = (EditText) findViewById(R.id.txt1);
        mPassCodeEditTexts[2] = (EditText) findViewById(R.id.txt2);
        mPassCodeEditTexts[3] = (EditText) findViewById(R.id.txt3);

        if (ACTION_REQUEST.equals(getIntent().getAction())) {
            /// this is a pass code request; the user has to input the right value
            mPassCodeHdr.setText(R.string.pass_code_enter_pass_code);
            mPassCodeHdrExplanation.setVisibility(View.INVISIBLE);
            setCancelButtonEnabled(false);      // no option to cancel

        } else if (ACTION_ENABLE.equals(getIntent().getAction())) {
            if (savedInstanceState != null) {
                mConfirmingPassCode = savedInstanceState.getBoolean(PassCodeActivity.KEY_CONFIRMING_PASSCODE);
                mPassCodeDigits = savedInstanceState.getStringArray(PassCodeActivity.KEY_PASSCODE_DIGITS);
            }
            if(mConfirmingPassCode){
                //the app was in the passcodeconfirmation
                requestPassCodeConfirmation();
            }else{
                /// pass code preference has just been activated in Preferences;
                // will receive and confirm pass code value
                mPassCodeHdr.setText(R.string.pass_code_configure_your_pass_code);
                //mPassCodeHdr.setText(R.string.pass_code_enter_pass_code);
                // TODO choose a header, check iOS
                mPassCodeHdrExplanation.setVisibility(View.VISIBLE);
                setCancelButtonEnabled(true);
            }

        } else if (ACTION_DISABLE.equals(getIntent().getAction())) {
            /// pass code preference has just been disabled in Preferences;
            // will confirm user knows pass code, then remove it
            mPassCodeHdr.setText(R.string.pass_code_remove_your_pass_code);
            mPassCodeHdrExplanation.setVisibility(View.INVISIBLE);
            setCancelButtonEnabled(true);

        } else {
            throw new IllegalArgumentException("A valid ACTION is needed in the Intent passed to "
                    + TAG);
        }

        setTextListeners();
    }


    /**
     * Enables or disables the cancel button to allow the user interrupt the ACTION
     * requested to the activity.
     *
     * @param enabled       'True' makes the cancel button available, 'false' hides it.
     */
    protected void setCancelButtonEnabled(boolean enabled){
        if(enabled){
            mBCancel.setVisibility(View.VISIBLE);
            mBCancel.setOnClickListener(new OnClickListener() {
                @Override
                public void onClick(View v) {
                    revertActionAndExit();
                }
            });
        } else {
            mBCancel.setVisibility(View.GONE);
            mBCancel.setVisibility(View.INVISIBLE);
            mBCancel.setOnClickListener(null);
        }
    }


    /**
     * Binds the appropiate listeners to the input boxes receiving each digit of the pass code.
     */
    protected void setTextListeners() {
    
         ///  First input field
        mPassCodeEditTexts[0].addTextChangedListener(new PassCodeDigitTextWatcher(0, false));


        /*------------------------------------------------
         *  SECOND BOX 
         -------------------------------------------------*/
        mPassCodeEditTexts[1].addTextChangedListener(new PassCodeDigitTextWatcher(1, false));

        mPassCodeEditTexts[1].setOnKeyListener(new View.OnKeyListener() {

            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if (keyCode == KeyEvent.KEYCODE_DEL && mBChange) {  // TODO WIP: event should be
                // used to control what's exactly happening with DEL, not any custom field...
                    mPassCodeEditTexts[0].setText("");
                    mPassCodeEditTexts[0].requestFocus();
                    if (!mConfirmingPassCode)
                        mPassCodeDigits[0] = "";
                    mBChange = false;

                } else if (!mBChange) {
                    mBChange = true;
                }
                return false;
            }
        });

        mPassCodeEditTexts[1].setOnFocusChangeListener(new View.OnFocusChangeListener() {

            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                /// TODO WIP: should take advantage of hasFocus to reduce processing
                if (mPassCodeEditTexts[0].getText().toString().equals("")) {  // TODO WIP validation
                // could be done in a global way, with a single OnFocusChangeListener for all the
                // input fields
                    mPassCodeEditTexts[0].requestFocus();
                }
            }
        });
        
        
        /*------------------------------------------------
         *  THIRD BOX
         -------------------------------------------------*/
        mPassCodeEditTexts[2].addTextChangedListener(new PassCodeDigitTextWatcher(2, false));

        mPassCodeEditTexts[2].setOnKeyListener(new View.OnKeyListener() {

            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if (keyCode == KeyEvent.KEYCODE_DEL && mBChange) {
                    mPassCodeEditTexts[1].requestFocus();
                    if (!mConfirmingPassCode)
                        mPassCodeDigits[1] = "";
                    mPassCodeEditTexts[1].setText("");
                    mBChange = false;

                } else if (!mBChange) {
                    mBChange = true;

                }
                return false;
            }
        });

        mPassCodeEditTexts[2].setOnFocusChangeListener(new View.OnFocusChangeListener() {

            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if (mPassCodeEditTexts[0].getText().toString().equals("")) {
                    mPassCodeEditTexts[0].requestFocus();
                } else if (mPassCodeEditTexts[1].getText().toString().equals("")) {
                    mPassCodeEditTexts[1].requestFocus();
                }
            }
        });


        /*------------------------------------------------
         *  FOURTH BOX
         -------------------------------------------------*/
        mPassCodeEditTexts[3].addTextChangedListener(new PassCodeDigitTextWatcher(3, true));

        mPassCodeEditTexts[3].setOnKeyListener(new View.OnKeyListener() {

            @Override
            public boolean onKey(View v, int keyCode, KeyEvent event) {
                if (keyCode == KeyEvent.KEYCODE_DEL && mBChange) {
                    mPassCodeEditTexts[2].requestFocus();
                    if (!mConfirmingPassCode)
                        mPassCodeDigits[2] = "";
                    mPassCodeEditTexts[2].setText("");
                    mBChange = false;

                } else if (!mBChange) {
                    mBChange = true;
                }
                return false;
            }
        });

        mPassCodeEditTexts[3].setOnFocusChangeListener(new View.OnFocusChangeListener() {

            @Override
            public void onFocusChange(View v, boolean hasFocus) {

                if (mPassCodeEditTexts[0].getText().toString().equals("")) {
                    mPassCodeEditTexts[0].requestFocus();
                } else if (mPassCodeEditTexts[1].getText().toString().equals("")) {
                    mPassCodeEditTexts[1].requestFocus();
                } else if (mPassCodeEditTexts[2].getText().toString().equals("")) {
                    mPassCodeEditTexts[2].requestFocus();
                }

            }
        });

    } // end setTextListener


    /**
     * Processes the pass code entered by the user just after the last digit was in.
     *
     * Takes into account the action requested to the activity, the currently saved pass code and
     * the previously typed pass code, if any.
     */
    private void processFullPassCode() {
        if (ACTION_REQUEST.equals(getIntent().getAction())) {
            if (checkPassCode()) {
                /// pass code accepted in request, user is allowed to access the app
                finish();

            }  else {
                showErrorAndRestart(R.string.pass_code_wrong, R.string.pass_code_enter_pass_code,
                        View.INVISIBLE);
            }

        } else if (ACTION_DISABLE.equals(getIntent().getAction())) {
            if (checkPassCode()) {
                /// pass code accepted when disabling, pass code is removed
                SharedPreferences.Editor appPrefs = PreferenceManager
                        .getDefaultSharedPreferences(getApplicationContext()).edit();
                appPrefs.putBoolean("set_pincode", false);  // TODO remove; this should be
                // unnecessary, was done before entering in the activity
                appPrefs.commit();

                Toast.makeText(PassCodeActivity.this, R.string.pass_code_removed, Toast.LENGTH_LONG).show();
                finish();

            } else {
                showErrorAndRestart(R.string.pass_code_wrong, R.string.pass_code_enter_pass_code,
                        View.INVISIBLE);
            }

        } else if (ACTION_ENABLE.equals(getIntent().getAction())) {
            /// enabling pass code
            if (!mConfirmingPassCode) {
                requestPassCodeConfirmation();

            } else if (confirmPassCode()) {
                /// confirmed: user typed the same pass code twice
                savePassCodeAndExit();

            } else {
                showErrorAndRestart(
                        R.string.pass_code_mismatch, R.string.pass_code_configure_your_pass_code, View.VISIBLE
                );
            }
        }
    }


    private void showErrorAndRestart(int errorMessage, int headerMessage,
                                     int explanationVisibility) {
        Arrays.fill(mPassCodeDigits, null);
        CharSequence errorSeq = getString(errorMessage);
        Toast.makeText(this, errorSeq, Toast.LENGTH_LONG).show();
        mPassCodeHdr.setText(headerMessage);                // TODO check if really needed
        mPassCodeHdrExplanation.setVisibility(explanationVisibility); // TODO check if really needed
        clearBoxes();
    }


    /**
     * Ask to the user for retyping the pass code just entered before saving it as the current pass
     * code.
     */
    protected void requestPassCodeConfirmation(){
        clearBoxes();
        mPassCodeHdr.setText(R.string.pass_code_reenter_your_pass_code);
        mPassCodeHdrExplanation.setVisibility(View.INVISIBLE);
        mConfirmingPassCode = true;
    }

    /**
     * Compares pass code entered by the user with the value currently saved in the app.
     *
     * @return     'True' if entered pass code equals to the saved one.
     */
    protected boolean checkPassCode(){
        SharedPreferences appPrefs = PreferenceManager
            .getDefaultSharedPreferences(getApplicationContext());

        String savedPassCodeDigits[] = new String[4];
        savedPassCodeDigits[0] = appPrefs.getString("PrefPinCode1", null);
        savedPassCodeDigits[1] = appPrefs.getString("PrefPinCode2", null);
        savedPassCodeDigits[2] = appPrefs.getString("PrefPinCode3", null);
        savedPassCodeDigits[3] = appPrefs.getString("PrefPinCode4", null);

        boolean result = true;
        for (int i = 0; i < mPassCodeDigits.length && result; i++) {
            result = result && (mPassCodeDigits[i] != null) &&
                    mPassCodeDigits[i].equals(savedPassCodeDigits[i]);
        }
        return result;
    }

    /**
     * Compares pass code retyped by the user in the input fields with the value entered just
     * before.
     *
     * @return     'True' if retyped pass code equals to the entered before.
     */
    protected boolean confirmPassCode(){
        mConfirmingPassCode = false;

        boolean result = true;
        for (int i = 0; i < mPassCodeEditTexts.length && result; i++) {
            result = result &&
                    ((mPassCodeEditTexts[i].getText().toString()).equals(mPassCodeDigits[i]));
        }
        return result;
    }

    /**
     * Sets the input fields to empty strings and puts the focus on the first one.
     */
    protected void clearBoxes(){
        for (int i=0; i < mPassCodeEditTexts.length; i++) {
            mPassCodeEditTexts[i].setText("");
        }
        mPassCodeEditTexts[0].requestFocus();
    }

    /**
     * Overrides click on the BACK arrow to correctly cancel ACTION_ENABLE or ACTION_DISABLE, while
     * preventing than ACTION_REQUEST may be worked around.
     *
     * @param keyCode       Key code of the key that triggered the down event.
     * @param event         Event triggered.
     * @return              'True' when the key event was processed by this method.
     */
    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event){
        if (keyCode == KeyEvent.KEYCODE_BACK && event.getRepeatCount()== 0){
            if (ACTION_ENABLE.equals(getIntent().getAction()) ||
                    ACTION_DISABLE.equals(getIntent().getAction())) {
                revertActionAndExit();
            }
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    /**
     * Saves the pass code input by the user as the current pass code.
     */
    protected void savePassCodeAndExit() {
        SharedPreferences.Editor appPrefs = PreferenceManager
                .getDefaultSharedPreferences(getApplicationContext()).edit();
        
        appPrefs.putString("PrefPinCode1", mPassCodeDigits[0]);
        appPrefs.putString("PrefPinCode2", mPassCodeDigits[1]);
        appPrefs.putString("PrefPinCode3", mPassCodeDigits[2]);
        appPrefs.putString("PrefPinCode4", mPassCodeDigits[3]);
        appPrefs.putBoolean("set_pincode", true);    /// TODO remove; unnecessary,
                                                     // Preferences did it before entering here
        appPrefs.commit();

        Toast.makeText(this, R.string.pass_code_stored, Toast.LENGTH_LONG).show();
        finish();
    }

    /**
     * Cancellation of ACTION_ENABLE or ACTION_DISABLE; reverts the enable or disable action done by
     * {@link Preferences}, then finishes.
     */
    protected void revertActionAndExit() {
        SharedPreferences.Editor appPrefsE = PreferenceManager
                .getDefaultSharedPreferences(getApplicationContext()).edit();

        SharedPreferences appPrefs = PreferenceManager
                .getDefaultSharedPreferences(getApplicationContext());

        boolean state = appPrefs.getBoolean("set_pincode", false);
        appPrefsE.putBoolean("set_pincode", !state);
        // TODO WIP: this is reverting the value of the preference because it was changed BEFORE
        // entering
        // TODO         in this activity; was the PreferenceCheckBox in the caller who did it
        appPrefsE.commit();
        finish();
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean(PassCodeActivity.KEY_CONFIRMING_PASSCODE, mConfirmingPassCode);
        outState.putStringArray(PassCodeActivity.KEY_PASSCODE_DIGITS, mPassCodeDigits);
    }


    private class PassCodeDigitTextWatcher implements TextWatcher {

        private int mIndex = -1;
        private boolean mLastOne = false;

        /**
         * Constructor
         *
         * @param index         Position in the pass code of the input field that will be bound to
         *                      this watcher.
         * @param lastOne       'True' means that watcher corresponds to the last position of the
         *                      pass code.
         */
        public PassCodeDigitTextWatcher(int index, boolean lastOne) {
            mIndex = index;
            mLastOne  = lastOne;
            if (mIndex < 0) {
                throw new IllegalArgumentException(
                        "Invalid index in " + PassCodeDigitTextWatcher.class.getSimpleName() +
                                " constructor"
                );
            }
        }

        private int next() {
            return mLastOne ? 0 : mIndex + 1;
        }

        /**
         * Performs several actions when the user types a digit in an input field:
         *  - saves the input digit to the state of the activity; this will allow retyping the
         *    pass code to confirm it.
         *  - moves the focus automatically to the next field
         *  - for the last field, triggers the processing of the full pass code
         *
         * @param s
         */
        @Override
        public void afterTextChanged(Editable s) {
            if (s.length() > 0) {
                if (!mConfirmingPassCode) {
                    mPassCodeDigits[mIndex] = mPassCodeEditTexts[mIndex].getText().toString();
                }
                mPassCodeEditTexts[next()].requestFocus();

                if (mLastOne) {
                    processFullPassCode();
                }

            } else {
                Log_OC.d(TAG, "Text box " + mIndex + " was cleaned");
            }
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            // nothing to do
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            // nothing to do
        }

    }


}

File: app/src/main/java/com/owncloud/android/ui/activity/PassCodeActivity.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2023 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2015 ownCloud Inc.
 * SPDX-FileCopyrightText: 2011 Bartek Przybylski
 * SPDX-License-Identifier: GPL-2.0-only AND (AGPL-3.0-or-later OR GPL-2.0-only)
 */
package com.owncloud.android.ui.activity

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.snackbar.Snackbar
import com.nextcloud.android.common.ui.theme.utils.ColorRole
import com.nextcloud.client.di.Injectable
import com.nextcloud.client.preferences.AppPreferences
import com.owncloud.android.R
import com.owncloud.android.authentication.PassCodeManager
import com.owncloud.android.databinding.PasscodelockBinding
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.ui.components.PassCodeEditText
import com.owncloud.android.utils.theme.ViewThemeUtils
import java.util.Arrays
import javax.inject.Inject

@Suppress("TooManyFunctions", "MagicNumber")
class PassCodeActivity : AppCompatActivity(), Injectable {

    companion object {
        private val TAG = PassCodeActivity::class.java.simpleName

        private const val KEY_PASSCODE_DIGITS = "PASSCODE_DIGITS"
        private const val KEY_CONFIRMING_PASSCODE = "CONFIRMING_PASSCODE"
        const val ACTION_REQUEST_WITH_RESULT = "ACTION_REQUEST_WITH_RESULT"
        const val ACTION_CHECK_WITH_RESULT = "ACTION_CHECK_WITH_RESULT"
        const val ACTION_CHECK = "ACTION_CHECK"
        const val KEY_PASSCODE = "KEY_PASSCODE"
        const val KEY_CHECK_RESULT = "KEY_CHECK_RESULT"
        const val PREFERENCE_PASSCODE_D = "PrefPinCode"
        const val PREFERENCE_PASSCODE_D1 = "PrefPinCode1"
        const val PREFERENCE_PASSCODE_D2 = "PrefPinCode2"
        const val PREFERENCE_PASSCODE_D3 = "PrefPinCode3"
        const val PREFERENCE_PASSCODE_D4 = "PrefPinCode4"
    }

    @JvmField
    @Inject
    var preferences: AppPreferences? = null

    @JvmField
    @Inject
    var passCodeManager: PassCodeManager? = null

    @JvmField
    @Inject
    var viewThemeUtils: ViewThemeUtils? = null

    @get:VisibleForTesting
    lateinit var binding: PasscodelockBinding
        private set

    private val passCodeEditTexts = arrayOfNulls<PassCodeEditText>(4)
    private var passCodeDigits: Array<String?>? = arrayOf("", "", "", "")
    private var confirmingPassCode = false
    private var changed = true // to control that only one blocks jump

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = PasscodelockBinding.inflate(layoutInflater)
        setContentView(binding.root)

        applyTint()
        setupPasscodeEditTexts()
        setSoftInputMode()
        setupUI(savedInstanceState)
        setTextListeners()
    }

    private fun applyTint() {
        viewThemeUtils?.platform?.colorViewBackground(binding.cardViewContent, ColorRole.SURFACE_VARIANT)
        viewThemeUtils?.material?.colorMaterialButtonPrimaryBorderless(binding.cancel)
    }

    private fun setupPasscodeEditTexts() {
        passCodeEditTexts[0] = binding.txt0
        passCodeEditTexts[1] = binding.txt1
        passCodeEditTexts[2] = binding.txt2
        passCodeEditTexts[3] = binding.txt3

        passCodeEditTexts.forEach {
            it?.let { viewThemeUtils?.platform?.colorEditText(it) }
        }

        passCodeEditTexts[0]?.requestFocus()
    }

    private fun setSoftInputMode() {
        window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_VISIBLE)
    }

    private fun setupUI(savedInstanceState: Bundle?) {
        if (ACTION_CHECK == intent.action) {
            // / this is a pass code request; the user has to input the right value
            binding.header.setText(R.string.pass_code_enter_pass_code)
            binding.explanation.visibility = View.INVISIBLE
            setCancelButtonEnabled(false) // no option to cancel
            showDelay()
        } else if (ACTION_REQUEST_WITH_RESULT == intent.action) {
            if (savedInstanceState != null) {
                confirmingPassCode = savedInstanceState.getBoolean(KEY_CONFIRMING_PASSCODE)
                passCodeDigits = savedInstanceState.getStringArray(KEY_PASSCODE_DIGITS)
            }
            if (confirmingPassCode) {
                // the app was in the passcode confirmation
                requestPassCodeConfirmation()
            } else {
                // pass code preference has just been activated in SettingsActivity;
                // will receive and confirm pass code value
                binding.header.setText(R.string.pass_code_configure_your_pass_code)
                binding.explanation.visibility = View.VISIBLE
            }
            setCancelButtonEnabled(true)
        } else if (ACTION_CHECK_WITH_RESULT == intent.action) {
            // pass code preference has just been disabled in SettingsActivity;
            // will confirm user knows pass code, then remove it
            binding.header.setText(R.string.pass_code_remove_your_pass_code)
            binding.explanation.visibility = View.INVISIBLE
            setCancelButtonEnabled(true)
        } else {
            throw IllegalArgumentException("A valid ACTION is needed in the Intent passed to $TAG")
        }
    }

    private fun setCancelButtonEnabled(enabled: Boolean) {
        binding.cancel.visibility = if (enabled) {
            View.VISIBLE
        } else {
            View.INVISIBLE
        }
        binding.cancel.setOnClickListener {
            if (enabled) {
                finish()
            }
        }
    }

    private fun setTextListeners() {
        for (i in passCodeEditTexts.indices) {
            val editText = passCodeEditTexts[i]
            val isLast = (i == 3)

            editText?.addTextChangedListener(PassCodeDigitTextWatcher(i, isLast))

            if (i > 0) {
                setOnKeyListener(i)
            }

            editText?.onFocusChangeListener = View.OnFocusChangeListener { _: View?, _: Boolean ->
                onPassCodeEditTextFocusChange(i)
            }
        }
    }

    private fun onPassCodeEditTextFocusChange(passCodeIndex: Int) {
        for (i in 0 until passCodeIndex) {
            if (TextUtils.isEmpty(passCodeEditTexts[i]?.text)) {
                passCodeEditTexts[i]?.requestFocus()
                break
            }
        }
    }

    private fun setOnKeyListener(passCodeIndex: Int) {
        passCodeEditTexts[passCodeIndex]?.setOnKeyListener { _: View?, keyCode: Int, _: KeyEvent? ->
            if (keyCode == KeyEvent.KEYCODE_DEL && changed) {
                passCodeEditTexts[passCodeIndex - 1]?.requestFocus()

                if (!confirmingPassCode) {
                    passCodeDigits?.set(passCodeIndex - 1, "")
                }

                passCodeEditTexts[passCodeIndex - 1]?.setText(R.string.empty)

                changed = false
            } else if (!changed) {
                changed = true
            }
            false
        }
    }

    /**
     * Processes the pass code entered by the user just after the last digit was in.
     *
     *
     * Takes into account the action requested to the activity, the currently saved pass code and the previously typed
     * pass code, if any.
     */
    private fun processFullPassCode() {
        if (ACTION_CHECK == intent.action) {
            if (checkPassCode()) {
                preferences?.resetPinWrongAttempts()

                // / pass code accepted in request, user is allowed to access the app
                passCodeManager?.updateLockTimestamp()
                hideSoftKeyboard()
                finish()
            } else {
                preferences?.increasePinWrongAttempts()
                showErrorAndRestart(R.string.pass_code_wrong, R.string.pass_code_enter_pass_code, View.INVISIBLE)
            }
        } else if (ACTION_CHECK_WITH_RESULT == intent.action) {
            if (checkPassCode()) {
                passCodeManager?.updateLockTimestamp()

                val resultIntent = Intent()
                resultIntent.putExtra(KEY_CHECK_RESULT, true)
                setResult(RESULT_OK, resultIntent)
                hideSoftKeyboard()
                finish()
            } else {
                showErrorAndRestart(R.string.pass_code_wrong, R.string.pass_code_enter_pass_code, View.INVISIBLE)
            }
        } else if (ACTION_REQUEST_WITH_RESULT == intent.action) {
            // / enabling pass code
            if (!confirmingPassCode) {
                requestPassCodeConfirmation()
            } else if (confirmPassCode()) {
                // / confirmed: user typed the same pass code twice
                savePassCodeAndExit()
            } else {
                showErrorAndRestart(
                    R.string.pass_code_mismatch,
                    R.string.pass_code_configure_your_pass_code,
                    View.VISIBLE
                )
            }
        }
    }

    private fun hideSoftKeyboard() {
        currentFocus?.let {
            val inputMethodManager = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodManager.hideSoftInputFromWindow(
                it.windowToken,
                0
            )
        }
    }

    private fun showErrorAndRestart(errorMessage: Int, headerMessage: Int, explanationVisibility: Int) {
        passCodeDigits?.let { Arrays.fill(it, null) }

        Snackbar.make(findViewById(android.R.id.content), getString(errorMessage), Snackbar.LENGTH_LONG).show()
        binding.header.setText(headerMessage) // TODO check if really needed
        binding.explanation.visibility = explanationVisibility // TODO check if really needed
        clearBoxes()
        showDelay()
    }

    /**
     * Ask to the user for retyping the pass code just entered before saving it as the current pass code.
     */
    private fun requestPassCodeConfirmation() {
        clearBoxes()
        binding.header.setText(R.string.pass_code_reenter_your_pass_code)
        binding.explanation.visibility = View.INVISIBLE
        confirmingPassCode = true
    }

    private fun checkPassCode(): Boolean {
        val savedPassCodeDigits = preferences?.passCode
        return passCodeDigits?.zip(savedPassCodeDigits.orEmpty()) { input, saved ->
            input != null && input == saved
        }?.all { it } ?: false
    }

    private fun confirmPassCode(): Boolean {
        return passCodeEditTexts.indices.all { i ->
            passCodeEditTexts[i]?.text.toString() == passCodeDigits!![i]
        }
    }

    private fun clearBoxes() {
        passCodeEditTexts.forEach { it?.text?.clear() }
        passCodeEditTexts.firstOrNull()?.requestFocus()
    }

    /**
     * Overrides click on the BACK arrow to correctly cancel ACTION_ENABLE or ACTION_DISABLE, while preventing than
     * ACTION_CHECK may be worked around.
     *
     * @param keyCode Key code of the key that triggered the down event.
     * @param event   Event triggered.
     * @return 'True' when the key event was processed by this method.
     */
    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && event.repeatCount == 0) {
            if (ACTION_CHECK == intent.action) {
                moveTaskToBack(true)
                finishAndRemoveTask()
            } else if (ACTION_REQUEST_WITH_RESULT == intent.action || ACTION_CHECK_WITH_RESULT == intent.action) {
                finish()
            } // else, do nothing, but report that the key was consumed to stay alive
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    private fun savePassCodeAndExit() {
        val resultIntent = Intent()
        resultIntent.putExtra(
            KEY_PASSCODE,
            passCodeDigits!![0] + passCodeDigits!![1] + passCodeDigits!![2] + passCodeDigits!![3]
        )
        setResult(RESULT_OK, resultIntent)
        passCodeManager?.updateLockTimestamp()
        finish()
    }

    private fun showDelay() {
        val delay = preferences?.pinBruteForceDelay() ?: 0

        if (delay <= 0) {
            return
        }

        binding.explanation.setText(R.string.brute_force_delay)
        binding.explanation.visibility = View.VISIBLE
        binding.txt0.isEnabled = false
        binding.txt1.isEnabled = false
        binding.txt2.isEnabled = false
        binding.txt3.isEnabled = false

        Thread(object : Runnable {
            override fun run() {
                try {
                    Thread.sleep(delay * 1000L)

                    runOnUiThread {
                        binding.explanation.visibility = View.INVISIBLE
                        binding.txt0.isEnabled = true
                        binding.txt1.isEnabled = true
                        binding.txt2.isEnabled = true
                        binding.txt3.isEnabled = true
                    }
                } catch (e: InterruptedException) {
                    Log_OC.e(this, "Could not delay password input prompt")
                }
            }
        }).start()
    }

    public override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(KEY_CONFIRMING_PASSCODE, confirmingPassCode)
        outState.putStringArray(KEY_PASSCODE_DIGITS, passCodeDigits)
    }

    private inner class PassCodeDigitTextWatcher(index: Int, lastOne: Boolean) : TextWatcher {
        private var mIndex = -1
        private val mLastOne: Boolean

        init {
            mIndex = index
            mLastOne = lastOne

            require(mIndex >= 0) {
                "Invalid index in " + PassCodeDigitTextWatcher::class.java.simpleName +
                    " constructor"
            }
        }

        private operator fun next(): Int {
            return if (mLastOne) 0 else mIndex + 1
        }

        /**
         * Performs several actions when the user types a digit in an input field: - saves the input digit to the state
         * of the activity; this will allow retyping the pass code to confirm it. - moves the focus automatically to the
         * next field - for the last field, triggers the processing of the full pass code
         *
         * @param s Changed text
         */
        override fun afterTextChanged(s: Editable) {
            if (s.isNotEmpty()) {
                if (!confirmingPassCode) {
                    val passCodeText = passCodeEditTexts[mIndex]?.text

                    if (passCodeText != null) {
                        passCodeDigits!![mIndex] = passCodeText.toString()
                    }
                }

                if (mLastOne) {
                    processFullPassCode()
                } else {
                    passCodeEditTexts[next()]?.requestFocus()
                }
            } else {
                Log_OC.d(TAG, "Text box $mIndex was cleaned")
            }
        }

        @Suppress("EmptyFunctionBlock")
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
        }

        @Suppress("EmptyFunctionBlock")
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        }
    }
}
